### ISSUE TITLE
cwltool relocateOutputs 'move' action for type Directory output

### ISSUE THREAD

**Issue by carcason** [common-workflow-language/cwltool#910]
## Expected Behavior
 relocateFunction moves src directory into dst directoy

## Actual Behavior
 CRASH

## Workflow Code
```
cwlVersion: v1.0
class: ExpressionTool

requirements:
  - class: InlineJavascriptRequirement

id: "files_to_folder"
inputs:
  input_files:
    type:
    - File[]
    - File
outputs:
  folder: Directory
expression: |
  ${
    var folder = {
      "class": "Directory",
      "basename": "",
      "listing": []
    }
    var files = [];
    if (!Array.isArray(inputs.input_files)){
      files.push (inputs.input_files)
    } else {
      files = inputs.input_files
    }
    for (var i = 0; i < files.length; i++){
      if (files[i].secondaryFiles && files[i].secondaryFiles.length > 0){
        Array.prototype.push.apply(folder.listing, files[i].secondaryFiles);
        delete files[i].secondaryFiles;
      }
      folder.listing.push (files[i])
    }
    return { "folder": folder };
  }

doc: |
  Returns Directory object with a listing set from File or File[]. Only one nested level of secondaryFiles
  is supported
```

## Full Traceback
```pytb
[2018-09-19 09:43:43,806] {base_task_runner.py:107} INFO - Job 138: Subtask CWLJobGatherer [2018-09-19 09:43:43,806] {cli.py:492} INFO - Running <TaskInstance: bowtie-index.CWLJobGatherer 2018-09-19T0
[2018-09-19 09:43:43,826] {models.py:1736} ERROR - 'posix.DirEntry' object has no attribute 'startswith'                                                                                                
Traceback (most recent call last):                                                                                                                                                                      
  File "/usr/local/lib/python3.6/site-packages/airflow/models.py", line 1633, in _run_raw_task                                                                                                          
    result = task_copy.execute(context=context)                                                                                                                                                         
  File "/usr/local/lib/python3.6/site-packages/cwl_airflow_parser/operators/cwljobgatherer.py", line 109, in execute                                                                                    
    return self.cwl_gather(context)                                                                                                                                                                     
  File "/usr/local/lib/python3.6/site-packages/cwl_airflow_parser/operators/cwljobgatherer.py", line 92, in cwl_gather                                                                                  
    StdFsAccess(""))                                                                                                                                                                                    
  File "/usr/local/lib/python3.6/site-packages/cwltool/process.py", line 327, in relocateOutputs                                                                                                        
    stageFiles(pm, stageFunc=_relocate, symLink=False)                                                                                                                                                  
  File "/usr/local/lib/python3.6/site-packages/cwltool/process.py", line 249, in stageFiles                                                                                                             
    stageFunc(p.resolved, p.target)                                                                                                                                                                     
  File "/usr/local/lib/python3.6/site-packages/cwltool/process.py", line 310, in _relocate                                                                                                              
    _relocate(dir_entry, os.path.join(dst, dir_entry.name))                                                                                                                                             
  File "/usr/local/lib/python3.6/site-packages/cwltool/process.py", line 305, in _relocate                                                                                                              
    if any(src.startswith(path + "/") for path in source_directories):                                                                                                                                  
  File "/usr/local/lib/python3.6/site-packages/cwltool/process.py", line 305, in <genexpr>                                                                                                              
    if any(src.startswith(path + "/") for path in source_directories):                                                                                                                                  
AttributeError: 'posix.DirEntry' object has no attribute 'startswith'                                                                                                                                   
[2018-09-19 09:43:43,828] {models.py:1756} INFO - Marking task as UP_FOR_RETRY       
```

## Your Environment
* cwltool version: 1.0.20180912090223
Check using ``cwltool --version``

## Code with issue proccess.py line 302

```
        if action == "move":
            # do not move anything if we are trying to move an entity from
            # outside of the source directories
            if any(src.startswith(path + "/") for path in source_directories):
                _logger.debug("Moving %s to %s", src, dst)
                if os.path.isdir(src) and os.path.isdir(dst):
                    # merge directories
                    for dir_entry in scandir(src):
                        _relocate(dir_entry, os.path.join(dst, dir_entry.name))
                else:
                    shutil.move(src, dst)
                    return

```

**Comment by psafont** [comment]
I'm unable to reproduce the crash with a testcase I created:

```json
{"input_files": [
  {
    "class": "File",
    "location": "910.json",
    "secondaryFiles": [
      {
        "class": "Directory",
        "location": ".."
      }
    ]
  },
  {
    "class": "File",
    "location": "910.cwl",
    "secondaryFiles": [
      {
        "class": "Directory",
        "location": "."
      }
    ]
  }
]}

```

Could you also provide the input for the job as well as the command line used to launch?

**Comment by michael-kotliar** [comment]
Here is an example to replicate the bug
```yaml
cwlVersion: v1.0
class: CommandLineTool

requirements:
- class: InlineJavascriptRequirement
- class: InitialWorkDirRequirement
  listing: |
    ${
      return [{"class": "Directory",
               "basename": "default",
               "listing": [{"class": "File",
                            "basename": "file.txt",
                            "contents": "Hello world"}
                           ],
               "writable": true}]
    }

hints:
- class: DockerRequirement
  dockerPull: ubuntu

inputs: []

outputs:

  output_folder:
    type: Directory
    outputBinding:
      glob: "*"

baseCommand: [ls]
```
When you run it twice, the second run gives error:
```
Moving /private/tmp/docker_tmpxvwge62q/default to /Users/kot4or/temp/del/default
Unhandled error:
  'posix.DirEntry' object has no attribute 'startswith'
```
It happens if output directory `default` already exist. If you delete it or set `"listing": []` in CWL file, everything works fine.

**Comment by psafont** [comment]
Thanks a lot!

If I fix the `_relocate` invocation in line 310 with `dir_entry.path` the second invocation fails the checksum because file.txt gets deleted, more errors uncovered:

```
[job 910.cwl] Removing input staging directory /tmp/tmp2m87tuh6
[job 910.cwl] Removing temporary directory /tmp/tmpl6je5y5k
Moving /tmp/tmpn211vn75/default to /home/psafont/dev/cwl/cwltool/default
Moving /tmp/tmpn211vn75/default/file.txt to /home/psafont/dev/cwl/cwltool/default/file.txt
Copying /tmp/tmpn211vn75/default to /home/psafont/dev/cwl/cwltool/default
Unhandled error:
  [Errno 2] No such file or directory: '/home/psafont/dev/cwl/cwltool/default/file.txt'
Traceback (most recent call last):
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/main.py", line 712, in main
    logger=_logger)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/executors.py", line 40, in __call__
    return self.execute(*args, **kwargs)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/executors.py", line 97, in execute
    path_mapper=runtime_context.path_mapper)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/process.py", line 337, in relocateOutputs
    visit_class(outputObj, ("File",), functools.partial(compute_checksums, fs_access))
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/utils.py", line 210, in visit_class
    visit_class(rec[d], cls, op)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/utils.py", line 210, in visit_class
    visit_class(rec[d], cls, op)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/utils.py", line 213, in visit_class
    visit_class(d, cls, op)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/utils.py", line 208, in visit_class
    op(rec)
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/process.py", line 974, in compute_checksums
    with fs_access.open(fileobj["location"], "rb") as f:
  File "/home/psafont/.local/venvs/cwltool/lib/python3.7/site-packages/cwltool/stdfsaccess.py", line 41, in open
    return open(self._abs(fn), mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/psafont/dev/cwl/cwltool/default/file.txt'
```

**Comment by mr-c** [comment]
Fixed in #912 thanks @psafont !
